torchtune/modules/transforms/tokenizers/_hf_tokenizer.py

```diff
@@ -146,3 +149,108 @@ def decode(self, token_ids: List[int]) -> str:
             str: The decoded string.
         """
         return self.tokenizer.decode(token_ids)
+
+
+def _infer_special_tokens_from_hf_config(config: dict) -> list[str]:
+    special_tokens = set()
+
+    standard_keys = [
+        "bos_token",
+        "eos_token",
+        "pad_token",
+        "unk_token",
+        "sep_token",
+        "cls_token",
+        "mask_token",
+    ]
+
+    for key in standard_keys:
+        if token := config.get(key):
+            if isinstance(token, str):
+                content = token
+            else:
+                content = token.get("content")
+
+            if content:
+                special_tokens.add(content)
+
+    for token in config.get("additional_special_tokens", []):
+        if isinstance(token, str):
+            content = token
+        else:
+            content = token.get("content")
+
+        if content:
+            special_tokens.add(content)
+
+    for token_info in config.get("added_tokens_decoder", {}).values():
+        if token_info.get("special", False):
+            if content := token_info.get("content"):
+                special_tokens.add(content)
+
+    # We sort lexicographically in order to get real tokens after all <|dummy_x|>
+    return sorted(list(special_tokens))
+
+
+class HuggingFaceModelTokenizer(ModelTokenizer):
+    def __init__(
+        self,
+        tokenizer_json_path: str,
+        *,
+        tokenizer_config_json_path: Optional[str] = None,
+        generation_config_path: Optional[str] = None,
+    ):
+        self.base_tokenizer = HuggingFaceBaseTokenizer(
+            tokenizer_json_path=tokenizer_json_path,
+            tokenizer_config_json_path=tokenizer_config_json_path,
+            generation_config_path=generation_config_path,
+        )
```

Comment:
I know @joecummings had some thoughts on whether we should use generic `base_tokenizer` instead of constraining to use `HuggingFaceBaseTokenizer`. I suspect the latter is better for making sure everything works together, but I know at least `Qwen2Tokenizer` still relies on the merges + vocab files instead of the `tokenizer.json` file (I alluded to this at the very bottom of #2706). So we should figure out if this will work for that case